datastore/transaction.go

```diff
@@ -299,12 +356,70 @@ func (t *Transaction) Rollback() (err error) {
 	if t.state == transactionStateExpired {
 		return errExpiredTransaction
 	}
-	t.state = transactionStateExpired
+
+	err = t.beginLaterTransaction()
+	if err != nil {
+		return err
+	}
+
 	_, err = t.client.client.Rollback(t.ctx, &pb.RollbackRequest{
 		ProjectId:   t.client.dataset,
 		DatabaseId:  t.client.databaseID,
 		Transaction: t.id,
 	})
+	if err != nil {
+		return err
+	}
+
+	t.state = transactionStateExpired
+	return nil
+}
+
+func (t *Transaction) parseReadOptions() (*pb.ReadOptions, error) {
+	var opts *pb.ReadOptions
+	switch t.state {
+	case transactionStateExpired:
+		return nil, errExpiredTransaction
+	case transactionStateInProgress:
+		opts = &pb.ReadOptions{
+			// Use existing transaction id for this request
+			ConsistencyType: &pb.ReadOptions_Transaction{Transaction: t.id},
+		}
+	case transactionStateNotStarted:
+		tOptionsPb, _ := t.parseTransactionOptions()
+		opts = &pb.ReadOptions{
+			// Begin a new transaction for this request
+			ConsistencyType: &pb.ReadOptions_NewTransaction{NewTransaction: tOptionsPb},
+		}
+	}
+	return opts, nil
+}
+
+func (t *Transaction) get(spanName string, keys []*Key, dst interface{}) (err error) {
+	t.ctx = trace.StartSpan(t.ctx, spanName)
+	defer func() { trace.EndSpan(t.ctx, err) }()
+
+	if t.state == transactionStateNotStarted {
```

Comment:
It looks like this business logic also exists in `nextBatch`. Maybe the `defer` gets in the way of being able to move this into a common function, but is there any way for this mutex locking logic to live in only one place? Perhaps this can be done using a decorator pattern?

Reply:
I couldn't find a way to implement decorator pattern but used [this](https://stackoverflow.com/a/68935422/2419636) as reference to move defer to new function.